File: source/go/json_config.go
package main

import (
	"encoding/json"
	"fmt"
)

type Config struct {
	Host           string `json:"host"`
	Port           int    `json:"port"`
	Debug          bool   `json:"debug"`
	LogLevel       string `json:"log_level"`
	MaxConnections int    `json:"max_connections"`
	Timeout        int    `json:"timeout"`
}

func prettyPrint(cfg Config) {
	data, _ := json.MarshalIndent(cfg, "", "  ")
	fmt.Println(string(data))
}

func mergeConfig(base Config, overrides map[string]interface{}) Config {
	result := base
	if v, ok := overrides["port"]; ok {
		result.Port = int(v.(float64))
	}
	if v, ok := overrides["host"]; ok {
		result.Host = v.(string)
	}
	if v, ok := overrides["debug"]; ok {
		result.Debug = v.(bool)
	}
	if v, ok := overrides["log_level"]; ok {
		result.LogLevel = v.(string)
	}
	if v, ok := overrides["max_connections"]; ok {
		result.MaxConnections = int(v.(float64))
	}
	if v, ok := overrides["timeout"]; ok {
		result.Timeout = int(v.(float64))
	}
	return result
}

func main() {
	defaults := Config{
		Host:           "localhost",
		Port:           8080,
		Debug:          false,
		LogLevel:       "info",
		MaxConnections: 100,
		Timeout:        30,
	}

	userConfigStr := `{"port": 3000, "debug": true, "log_level": "debug"}`
	var userOverrides map[string]interface{}
	json.Unmarshal([]byte(userConfigStr), &userOverrides)

	finalConfig := mergeConfig(defaults, userOverrides)

	fmt.Println("=== Default Config ===")
	prettyPrint(defaults)
	fmt.Println("=== User Overrides ===")
	userCfgData, _ := json.MarshalIndent(userOverrides, "", "  ")
	fmt.Println(string(userCfgData))
	fmt.Println("=== Final Config ===")
	prettyPrint(finalConfig)
}

File: source/rust/json_config.rs
// Simple JSON-like config without external crates

#[derive(Clone, Debug)]
struct Config {
    host: String,
    port: i32,
    debug: bool,
    log_level: String,
    max_connections: i32,
    timeout: i32,
}

impl Config {
    fn defaults() -> Self {
        Config {
            host: "localhost".to_string(),
            port: 8080,
            debug: false,
            log_level: "info".to_string(),
            max_connections: 100,
            timeout: 30,
        }
    }

    fn pretty_print(&self) {
        println!("{{");
        println!("  \"host\": \"{}\",", self.host);
        println!("  \"port\": {},", self.port);
        println!("  \"debug\": {},", self.debug);
        println!("  \"log_level\": \"{}\",", self.log_level);
        println!("  \"max_connections\": {},", self.max_connections);
        println!("  \"timeout\": {}", self.timeout);
        println!("}}");
    }
}

fn main() {
    let defaults = Config::defaults();

    // User overrides
    let mut final_config = defaults.clone();
    final_config.port = 3000;
    final_config.debug = true;
    final_config.log_level = "debug".to_string();

    println!("=== Default Config ===");
    defaults.pretty_print();
    println!("=== User Overrides ===");
    println!("{{");
    println!("  \"port\": 3000,");
    println!("  \"debug\": true,");
    println!("  \"log_level\": \"debug\"");
    println!("}}");
    println!("=== Final Config ===");
    final_config.pretty_print();
}
